R/extract_diagnostics.R

```diff
@@ -0,0 +1,161 @@
+#' Extract diagnostic metrics from model fit and data
+#'
+#' This function extracts various diagnostic metrics from a fitted `epinow2`
+#' model and provided data. It checks for low case counts and computes
+#' diagnostics from the fitted model, including the mean acceptance
+#' statistic, divergent transitions, maximum tree depth, and Rhat values.
+#' These diagnostics are then flagged if they exceed specific thresholds,
+#' and the results are returned as a data frame.
+#'
+#' @param fit A list containing the model fit object from `epinow2`, which
+#'            includes `estimates$fit`.
+#' @param data A data frame containing the input data used in the model fit.
+#' @param job_id A unique identifier for the job or task being processed.
+#' @param task_id A unique identifier for the task being performed.
+#' @param disease,geo_value,model Metadata for downstream processing.
+#'
+#' @return A \code{data.frame} containing the extracted diagnostic metrics. The
+#' data frame includes the following columns:
+#' \itemize{
+#'   \item \code{diagnostic}: The name of the diagnostic metric.
+#'   \item \code{value}: The value of the diagnostic metric.
+#'   \item \code{state}: The state for which the model was run.
+#'   \item \code{disease}: The disease/pathogen being analyzed.
+#'   \item \code{job_id}: The unique identifier for the job.
+#'   \item \code{task_id}: The unique identifier for the task.
+#'   \item \code{disease,geo_value,model}: Metadata for downstream processing.
+#' }
+#'
+#' @details
+#' The following diagnostics are calculated:
+#' \itemize{
+#'   \item \code{mean_accept_stat}: The average acceptance statistic across
+#'         all chains.
+#'   \item \code{p_divergent}: The proportion of divergent transitions across
+#'         all samples.
+#'   \item \code{p_max_treedepth}: The proportion of samples that hit the
+#'         maximum tree depth.
+#'   \item \code{p_high_rhat}: The proportion of parameters with Rhat values
```

Comment:
Would it be useful to have the number as well as the proportion?

Reply:
I'm not sure. I think proportion is more important because absolute number is a function of the number of samples. But open to exposing it if it would be helpful?

Reply:
I was talking about the parameters here so that'd be a function of number of parameters. But yeah same deal. I guess if someone tell me it's some % of parameters my first question is how many parameters are there total. Also which parameter is it

Reply:
Oh whoops yes you're correct. I'll add.

Reply:
Have you gotten to using these features in meetings yet? I guess it'd be good here to maybe link up to what the proposed solution to having this information in review meetings is

Reply:
So for example is there some roadmap to like having traceplots available on the dashboard? That'd be good. Of course having some summary diagnostics up there / available could be nice as a first step

Reply:
I would also really like those.

Katie and I have discussed this a few times. My current understanding is that a staffing plan for an updated dashboard for internal review is gated by this PR. As in, once we have a stable output schema planning on how to start updating internal monitoring will begin.

I think that dashboard would likely be light on the kinds of model diagnostics that we care about because it would be focused on the weekly review crowd as our audience. I was thinking about building an additional internal dashboard focused on model diagnostics. It would be a bit less friendly than what we'd expose to other centers in the weekly review meetings. But again, that's gated on my time freeing up from this pipeline to go work on that.

Reply:
I was also thinking that in #10 we could save the cmdstan CSVs as a pipeline output. That would make it easier to extract stan-specific diagnostics and feed those to a dashboard.

We should go discuss this over there, but I was weighing whether arviz made sense over bayesplot. I think arviz is a bit more language agnostic, so we could plug numpyro models into there as well.

Reply:
> I was thinking about building an additional internal dashboard focused on model diagnostics

r.e. this seems to me arguable that there can just be additional tabs or panels of dashboard clearly labelled as "developer" or "expert". I think not worth the effort to create 2 products here rather than just have tabs or panels. Even without the labelling people can just ignore things they don't understand (but ofc I think there should be clear labelling)

Reply:
Yes happy to move to an issue. To me #10 looks somewhat different to what we are talking about here, but I suppose the point is that we don't have a repo yet to make issues for this potential dashboard product and are still in the meetings phase about it. Happy to be included as optional on scoping meetings about the dashboard